cmake/CPM.cmake

```diff
@@ -505,17 +505,50 @@ function(cpm_override_fetchcontent contentName)
   set_property(GLOBAL PROPERTY ${propertyName} TRUE)
 endfunction()
 
+macro(cpm_cmake_eval)
+  set(__ARGN "${ARGN}")
+  if(COMMAND cmake_language)
+    cmake_language(EVAL CODE "${__ARGN}")
+  else()
+    file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/eval.cmake "${__ARGN}")
+    include(${CMAKE_CURRENT_BINARY_DIR}/eval.cmake)
+  endif()
+endmacro()
+
 # Download and add a package from source
-function(CPMAddPackage)
+macro(CPMAddPackage)
+  set(__ARGN "${ARGN}")
+  list(LENGTH __ARGN __ARGN_Length)
+  if(__ARGN_Length EQUAL 1)
+    cpm_add_package_single_arg(${ARGN})
+  else()
+    # Forward preserving empty string arguments
+    # (https://gitlab.kitware.com/cmake/cmake/-/merge_requests/4729)
+    set(__ARGN_Quoted)
+    foreach(__ARG IN LISTS __ARGN)
+      string(APPEND __ARGN_Quoted " [==[${__ARG}]==]")
+    endforeach()
+    cpm_cmake_eval("cpm_add_package_multi_arg( ${__ARGN_Quoted} )")
+  endif()
+endmacro()
```

Comment:
There is, however, a specific case to be aware of which can result in unexpected behavior. Because macros treat their arguments as string substitutions rather than as variables, if they use ARGN in a place where a variable name is expected, the variable it will refer to will be in the scope from which the macro is called, not the ARGN from the macro’s own arguments. 

From chapter 9.2 [professional-cmake](https://crascit.com/professional-cmake/)